# Determining Iris species through k-nn classification

### Project dependency

In [37]:
from sklearn import datasets
from sklearn.cross_validation import train_test_split
from sklearn.neighbors import KNeighborsClassifier

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
%matplotlib inline

In [5]:
plt.style.use('ggplot')

## Loading the Iris Dataset

In [6]:
iris = datasets.load_iris()

In [8]:
type(iris)

sklearn.datasets.base.Bunch

In [12]:
print (iris.keys())

dict_keys(['DESCR', 'target_names', 'data', 'feature_names', 'target'])


In [13]:
type(iris.data), type(iris.target)

(numpy.ndarray, numpy.ndarray)

In [16]:
iris.data.shape

(150, 4)

Our data consist of 4 columns and 150 rows. The values in each row describe the feature of a type of Iris. The corresponding species of Iris for each row is given in the target.

In [20]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], 
      dtype='<U10')

In [30]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Here, 0 stands of setosa, 1 for versicolor and 2 for virginica

In [18]:
iris.target.shape

(150,)

## Splitting the Training and test data

In [34]:
X = iris.data
y = iris.target

In [40]:
X_train, X_test, y_train, y_test= train_test_split(X,y, test_size=0.3 , random_state=25, stratify=y)

## k-nn Classification

In [39]:
knn= KNeighborsClassifier(n_neighbors = 6)

In [44]:
knn.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=6, p=2,
           weights='uniform')

### Predicting the target using test data 

In [42]:
y_pred = knn.predict(X_test)

In [43]:
y_pred

array([2, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 2, 1, 1, 0, 2, 2, 2, 1, 0, 1, 1, 2,
       2, 0, 1, 1, 0, 0, 1, 2, 2, 2, 2, 1, 1, 2, 0, 0, 2, 0, 1, 2, 1, 0])

In [45]:
y_test

array([2, 0, 1, 1, 1, 0, 2, 0, 0, 0, 0, 2, 1, 1, 0, 2, 2, 2, 1, 0, 1, 1, 2,
       2, 0, 1, 1, 0, 0, 1, 2, 2, 2, 2, 1, 1, 2, 0, 0, 2, 0, 1, 2, 1, 0])

We can see that the predicted target is similar to the actual target. This is because the data was fitted perfectly. Also, the dataset consists of information of only 150 species. With increasing no. of data the k-nn score will decrease

In [47]:
knn.score(X_test,y_pred)

1.0